hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java

```diff
@@ -520,20 +520,18 @@ public HFileScannerImpl(final HFile.Reader reader, final boolean cacheBlocks,
     }
 
     void updateCurrBlockRef(HFileBlock block) {
-      if (block != null && this.curBlock != null &&
-          block.getOffset() == this.curBlock.getOffset()) {
+      if (block != null && curBlock != null && block.getOffset() == curBlock.getOffset()) {
         return;
       }
-      // We don't have to keep ref to EXCLUSIVE type of block
-      if (this.curBlock != null && this.curBlock.usesSharedMemory()) {
```

Comment:
There is a  concern here. Even if the block is on an exclusive heap memory area, we will keep ref to that in this list.  In a Phoenix Aggregation kind of use case where many blocks might get fetched and not immediately shipped, we are keeping the ref unwantedly here for longer time. This makes the GC not able to reclaim the heap memory area for the blocks.  This might be a hidden bomb IMO.  Its not good to remove the MemType.  Lets create the block with memory type as EXCLUSIVE  when the block data is on heap. The block might be coming from LRU cache or by fetching the block data from HDFS into heap memory area.   When the block comes from off heap BC or if it is backed by a BB from the pool (While reading from HDFS, read into pooled BB) lets create the block with mem type as SHARED.   Every block can have the retain and release method but let the EXCLUSIVE types do a noop here.

Reply:
FIled an separate issue  for this: https://issues.apache.org/jira/browse/HBASE-21879.  Let finish that in an new one because I believe it's will be many changes , also can helop to make this PR forward.

Reply:
So already as part of this JIRA you removed the MEMTYPE and so even the bucket cache engines like FileIOEngine will remove the memtype and so all the blocks will go through the release() way by adding to prevBlocks and then getting released. The new PR #268 will remove the need for addition to prevBlocks as it is a noop. So now both bucket cache and read from HDFS both will create onheap/offheap blocks and the type of block will determine what it is .